x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -813,4 +855,96 @@ private static boolean isAsyncRelatedAction(String action) {
             || action.equals(EqlAsyncActionNames.EQL_ASYNC_GET_RESULT_ACTION_NAME)
             || action.equals(SqlAsyncActionNames.SQL_ASYNC_GET_RESULT_ACTION_NAME);
     }
+
+    /**
+     * A lazily loaded Set for authorized indices. It avoids loading the set if only contains check is required.
+     * It only loads the set if iterating through it is necessary, i.e. when expanding wildcards.
+     */
+    private static class AuthorizedIndicesSet implements Set<String> {
+
+        private final Supplier<Set<String>> supplier;
+        private final Predicate<String> predicate;
+        private Set<String> authorizedIndices = null;
+
+        AuthorizedIndicesSet(Supplier<Set<String>> supplier, Predicate<String> predicate) {
+            this.supplier = Objects.requireNonNull(supplier);
+            this.predicate = Objects.requireNonNull(predicate);
+        }
+
+        private Set<String> getAuthorizedIndices() {
+            if (authorizedIndices == null) {
+                authorizedIndices = supplier.get();
+            }
+            return authorizedIndices;
```

Comment:
I would **slightly** prefer we make this check thread safe, like in the `CachingAsyncSupplier`.
We're not using it in a multi-thread fashion, and I don't think we would ever do, and even if we did we only risk performance problems with building the set multiple times (but no correctness problems).
What do you think?

Reply:
This Set is wrapped in a `CachingAsyncSupplier`, i.e. `resolvedIndicesAsyncSupplier` (not `authorizedIndicesSupplier` which I will remove, see below). The `resolvedIndicesAsyncSupplier` is what gets passed to other places so that this Set is never directly used by anything else. Since `resolvedIndicesAsyncSupplier` already guarantees thread safety, I think the Set itself can be kept simple. I added the following comment to the class to clarify its usage:
```
* NOTE that the lazy loading does not guarantee to run only once and is not meant to be used by multi-threads
* because loading multiple times can incur performance penalty (but not correctness).
```

Reply:
I don't think this implementation can be considered thread safe. There's no memory barrier here, so there's no guarantee about the state of the `authorizedIndices` field when viewed from another thread.

The options are:
- simply state that it is not thread-safe and if callers require thread safety they need to synchronize
- mark `authorizedIndices` as volatile

Reply:
You are right. It is indeed not thread safe. I changed the documentation to state that and warned about changing how this class is used.